# Part 1 of week 3

# Instruction (copied from Coursera):

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

1. Start by creating a new Notebook for this assignment. (OK!)

2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas  dataframe like the one shown below:
<img src="https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/7JXaz3NNEeiMwApe4i-fLg_40e690ae0e927abda2d4bde7d94ed133_Screen-Shot-2018-06-18-at-7.17.57-PM.png?expiry=1614556800000&hmac=wEwibcmT8gvyPaPIsVVZleNL3OOUvQYRv4pItozNzn4"/>

3. To create the above dataframe:
- The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood __(OK! Step "c")__
- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned. __(OK! Step "e")__
- More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table.__(OK! Not necessary, as comented in step "f".)__
- If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough. __(OK! Step "g")__
- Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making. __(OK!)__
- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe. __(OK! Step "h")__

4. Submit a link to your Notebook on your Github repository.(10 marks) __(OK!)__

## a) Import the necessary libraries

In [55]:
import pandas as pd
import numpy as np

## b) Read the tables at the wikipedia url

In [37]:
url = 'http://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df_wiki1 = pd.read_html(url)
df_wiki1[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

## c) Create a data frame with the table we are interested

In [38]:
df_wiki2 = df_wiki1[0]
df_wiki2

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

## d) Replace the "Not assigned" boroughs to "NaN"

In [39]:
df_wiki2['Borough'].replace('Not assigned',np.nan, inplace=True)
df_wiki2

Postal Code           Borough  \
0           M1A               NaN   
1           M2A               NaN   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z               NaN   
176         M6Z               NaN   
177         M7Z               NaN   
178         M8Z         Etobicoke   
179         M9Z               NaN   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

## e) Drop the rows where the borough was 'Not assigned'

In [46]:
df_wiki2.dropna(axis=0,inplace=True)
df_wiki2

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## f) Verify if there is a repeated 'Postal Code'

In [54]:
df_wiki2.nunique()

Postal Code      103
Borough           10
Neighbourhood     99
dtype: int64

## Analysis:
#### There are 103 unique Postal Codes and 103 rows at the table. 
#### Thus, it's not necessary to do this step:
_"More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table."_

## g) Replace the Neighbourhood with 'Not assigned' with the name of the corresponding borough.

In [47]:
for i in range(df_wiki2.shape[0]):
    if df_wiki2.iloc[i][2] == 'Not assigned':
        df_wiki2.iloc[i][2] == df_wiki2.iloc[i][1]
df_wiki2        

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## h) In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe. 

In [58]:
df_wiki2.shape

(103, 3)